### PR TITLE
Make parse_command more idiomatic

### PR DESCRIPTION
**Pull request by Kixunil**
This refactors `parse_command` a bit to make it more idiomatic. The
changes are:

* Use of `?` to decrease nesting
* Removed `return`s
* Change `if let Some(foo) = bar { Some(baz) }` to `.map()` combinator
* Remove `{}` in `match` arms as those expressions are trivial (rustfmt)

(Noticed this during stream but it felt off-topic, so I made a PR now instead.)

### CODE REVIEW

File: src/editor.rs
Comment (kneasle) [inline src/editor.rs:270]: Could you add `#[rustfmt_skip]` to the function so that I don't reformat it when I save the file (I have format-on-save enabled to avoid loads of mess with commits that are only formatting).

**Review by kneasle — APPROVED** [review]
This definitely looks nicer!  I think at some point we should probably refactor out the duplicated functionality between `Action` and `Command` because adding a command required many changes to various parts of the code.  But I'm really not sure of the best way to deal with that and I'll leave refactoring until I'm happy that we'll get a good solution.

File: src/editor.rs
Comment (kneasle) [reply src/editor.rs:270]: Apparently you actually need `#[cfg_attr(rustfmt, rustfmt_skip)]` to make this work on stable.

Comment (Kixunil) [reply src/editor.rs:270]: This is how `rustfmt` formatted it, so it shouldn't change it anymore. Or am I missing something?

**Comment by Kixunil** [comment]
Oh, I think I know how to refactor the duplication, going to try it.

**Comment by Kixunil** [comment]
I attempted to refactor it by turning `Action` into `struct { command: Command, param: Option<char>, }`. Then I realized it'd lead to more error-prone code, throwing away some of Rusts advantages. I think macros will do the job better.

**Comment by Kixunil** [comment]
Hmm, it seems that the interaction between `Direction` and parameter makes this less easy then I expected. I will put this on hold until I have a bit more time to think about it.

File: src/editor.rs
Comment (kneasle) [reply src/editor.rs:270]: No you're quite right - this is how rustfmt formats it. I was very tired last night and completely forgot that the PR CI checks also run rustfmt and so this _had_ to be the way rusrfmt formatted it :laugh:.

**Comment by kneasle** [comment]
> I attempted to refactor it by turning `Action` into `struct { command: Command, param: Option<char>, }`

I also considered this when I was mulling over what to do with this, and I also couldn't immediately come up with a better way that was less code duplication but also didn't remove rust's safety. There's also the added complexity that commands can technically have arbitrary length (at least in Vim - e.g. `3da(` to delete 3 sets of nested brackets), and so our current nested match is gonna get pretty ugly once we throw things like counts into the mix.

The best solution I have so far is to model command and action simultaneously with a single struct with an interface that goes something like this, which feels quite like a pseudo-iterator:
```rust
#[derive(...)] 
enum CmdResult {
    Unfinished, 
    Invalid, 
    Complete, 
}

trait ParseCmd {
    /// produces a new empty command that uses a given KeyMap's key bindings
    fn new(keymap: &KeyMap) -> Self;

    /// consume a char, and return an enum value representing the state of the command
   fn consume_char(&mut self, c: char) -> CmdResult;

   /// once the command is complete, execute the command
   fn execute(self, &mut DAG) -> Result<... >;
} 
```

(also I've been playing with the idea of renaming 'Command' to 'Cmd' in the whole code base, because I'm pretty sure the standard library uses it as a shorthand and it feels more rusty)

This allows us to package the complexity of the command parsing into one single `struct`, and the code to drive this struct fits very nicely with the current event loop in `Editor`.

I guess `KeyMap` still needs to map to something, so we'd still need to have an enum of atomic commands, which would be like `Command` is now (where `Action` has now been replaced with `ParseCmd`.

What do you think?

**Comment by Kixunil** [comment]
I was actually also thinking about something like that and started writing it but I realized it wouldn't help significantly. The main issue is different actions can have different number of parameters. Which then allows for various invalid states to be represented.

One idea I was contemplating, but felt even worse is something like this:

```rust
enum Command {
    // Don't want to write them out, they are just bare enums
    NoParams(NoParamCommand),
    OneParam(OneParamCommand),
// more in the future?
}

enum ParsingState {
    IncompleteOneParam(OneParamCommand),
    CompleteNoParam(NoParamCommand),
    CompleteOneParam(OneParamCommand, char),
    Invalid,
}
```

As you can see, this adds a bunch of complexity and removes the duplication between `Command` and `Action`.
Maybe the complexity can be reduced with macros or some other trick. One more thing that makes macros troublesome is how `Direction` is in `Command`. If we flattened it instead, it would be easier to implement.

Rough idea of how the macro could look like:

```
command! {
    // no param
    Quit, 'q', "quit Sapling", Color::RED;
    // one param must also specify a format string with {}
    InsertChild, 'o', "insert child", Color::CYAN, "insert {} as last child";
    // ...
}
```

This would implement parsing, summary, default keymap (a nice property is you can't forget a key like you did on live stream :)), and `description_and_color`.

**Comment by kneasle** [comment]
I've got to admit that out of all the options mooted so far, I think the macro is by far the best, though to make it really epic it would be nice to able to pass in the identifier of the method of `DAG` that should be called by that command to completely remove the reliance on handwritten code.  But Rust's macro system is definitely flexible enough to handle that.

I do, however, have an extremely rogue idea.  There's an editor called [Kakoune](https://github.com/mawww/kakoune), which has a very nice twist/improvement on Vim's modal editing model.  In Kakoune, the 'grammar' of the normal-mode language is reversed, so that instead of the arguments for a meta-command coming **after** the command, the arguments instead **precede** the command.

As an example, the command `d3w` from Vim would be typed in Kakoune as `3Wd` (the change from `w` to `W` is a technicality of how Kakoune handles selections).  However, Kakoune executes this command in two parts (`3W` and `d`) as you type (so `3W` is a command that extends the selection by 3 words and `d` is a command with no arguments that simply deletes the current selection).  This has many advantages over Vim's model - mostly that you get feedback of what is about to be deleted, but also that `d` can have more than one sub-command as an argument, because something like `3Wld` is turned into `3W`, `l` and `d` which will have the effect of deleting 3 words and one character after the cursor.

One enormous advantage of this system is that (as far as I know) commands do not have arguments because their arguments only modify the selection before the command is applied.  Instead, commands always correspond to simple strings of characters, which can be parsed just by repeated string comparisons.  Therefore, the whole complexity of command parsing essentially disappears.

However, I don't want to alienate Vim users from Sapling by having Kakoune's grammar be baked into Sapling, but I do want it to be an option because it makes so much more sense.  Ideally, this change would be a configuration option which is in 'Vim mode' by default.

I therefore think that Sapling could 'compile' Vim-like commands into multiple Kakoune commands and then execute those (without saving the history in between edits).  There is still some necessary complexity from Vim's command structure, but this can probably offset by grouping commands into categories like `Movement`, `Insertion`, `Deletion`, `History`, etc. (these correspond to the colours I've been using in `description_and_color` since now the mapping between command categories and colors can be done at render-time which makes sure that only the `editor` module has to deal with details of how terminal rendering is handled).

I think we'd probably still want a macro to define the Vim-style command grammar but the meaning of the commands could be represented as simply an iterator over commands (or a  slice or Vec or anything else that implements `IntoIterator`).

What do you reckon?

**Comment by kneasle** [comment]
Thinking more about this, I forgot that commands like insert (`i`, `o` and `a`) and replace (`r`) both need to have arguments in Sapling but not in Kakoune.  Therefore, I revise my massive comment slightly because insertion and replacement both require arguments, but I don't think they can be put into meta-commands so the complexity is capped at having one argument.

**Comment by Protowalker** [comment]
Why not have two types of commands: `Unit` and `Stateful`
`Unit` would be something like array, `Stateful` would be `i` or `o` (having data)

**Comment by Kixunil** [comment]
>  it would be nice to able to pass in the identifier of the method of DAG that should be called by that command to completely remove the reliance on handwritten code. 

I was actually thinking about specifically **not** doing this because it felt too much. But now that I'm thinking about it, the `ParseCmd` trait above combined with macros could work well. The method would be just a trait method, so you wouldn't have to specify it in macro, just impl trait outside the macro. (I did something similar in a closed-source code I wrote once; it involved my `struct_deser` crate, which could be adapted for chars)

```rust
// Impls could be derived/generated
// for struct Output(A, B, C) it'd attempt to parse A, then B, then C
trait Parse: Default {
    type Output: Action;
    type Error: std::error::Error;
    // bool - true if char was consumed, false if not - e.g. to finish parsing of number
    fn consume_char(&mut self, c: char) -> Option<Result<(Self::Output, bool), Self::Error>>;
}

trait Parsable {
    type Parser: ParseAction<Output=Self>;
}

// replaces the current use of char
// use from primitive macro, parser is trivial ZST without state
enum JsonNodeKind {
    True,
    False,
    Null,
    String,
    Array,
    Object,
}

impl Parsable for JsonNodeKind {
// ...
}

// how many times a command should be repeated
struct Repeat(usize);

impl Parsable for Repeat {
// ...
}

trait Execute {
    fn execute(self, dag: &mut DAG);
}

action! {
    'q', "quit Sapling", Quit,
    'o', "insert child", InsertChild(Repeat, JsonNodeKind),
    'd', "delete node(s)", Delete(Repeat, Direction),
    // ...
}
```

**Comment by Kixunil** [comment]
Or rather something like this:

```rust
use displaydoc::Display;

/// quit Sapling
#[derive(Display, Parse)]
struct Quit;
/// insert child {1} to end {0} times
struct InsertChild(Repeat, JsonNodeKind),
// ...

action! {
'q', Quit;
'o', InsertChild;
// ...
}
```

Looks cleaner to me.

Also I realized, we probably want a newtype over char instead of `JsonNodeKind` to allow multiple languages more cleanly.

**Comment by kneasle** [comment]
I think we're definitely converging on a solution - at any rate, more ideas is always good.  Also, I don't think we need to get a perfect system before putting it into Sapling, since the whole project is an experiment and we can definitely do more refactoring later.  Things in Sapling do have to **work** though - they just don't have to start out being the most elegant solution.

I've been doing plenty of thinking about this subject, and I decided to come at this problem top-down (deciding exactly what we want the command parsing to do and then deciding how best to satisfy these requirements).

**I have an idea for how to structure this which I think fits the requirements, but I'm very open to other ideas.**

## Requirements

- Ultimately, the job of command parsing is to turn keyboard input into changes to the AST, so we need to essentially fill the `???` in the flow diagram with some number of phases:
  ```
           keyboard input (`&str` or `&[char]`)
  -[???]-> changes to the AST
  ```
- The keybindings of specific commands (`o`, `d`, `i`, etc.) need to be configurable.
- Because Sapling is all about getting better editing speed than conventional editors, we should be able to explore new ways of structuring commands than what Vim does.  But at the same time, I don't want Sapling to have an unnecessarily steep learning curve, so I think that by default Sapling should follow Vim's command structure as closely as possible.
- The code should be extensible and easy to modify/refactor
- The code should be as easy to reason about as possible

## Non-requirements
- Extremely good performance is not required - given the length of commands, this is unlikely to be a bottleneck.

## My proposed solution

My proposed solution is to build a compiler-style tokeniser-parser for the commands, corresponding to the following flow diagram:
 ```
                   keyboard input (`&str` or `&[char]`)
-[tokenisation]-> sequence of tokens (`Vec<CmdToken>`)
----[parsing]---> sequence of actions with counts (`Vec<(usize, Action)>`)
-----[apply]----> changes to the AST
```

At every step in the parsing, the command could be deduced to be either incomplete or invalid, in which case parsing should halt.  Therefore, each phase will return a `Result`-like type that looks something like:
```rust
enum ParseResult<T, E: std::error::Error> {
    /// This step couldn't find anything wrong with the user's input
    Ok(T),
    /// The user's input is in some way wrong and the command buffer should be cleared
    Invalid(E),
    /// The user's input is not valid yet, but we should keep the command buffer and await more keystrokes
    Incomplete
}
```

### Phase 1: Tokenisation

The first step of parsing a command is to apply the keymap to the keystrokes to just generate a stream of small command tokens (e.g. `d` -> `Delete`, `3` -> `Count(3)`, `oa` -> `InsertChild('a')`, etc.).  Longer strings are also possible, (e.g. in Vim we could have `gg` -> `GotoTopOfFile`).  This is very similar to tokenisation in general parsers, and I think corresponds roughly to the following specification:
```rust
enum CmdToken {
    Count(usize),
    Delete,
    InsertChild(char),
    ....
}

// `SomeError` is some error type that gives the caller more information about how the command is invalid
fn tokenise(command: &str, keymap: KeyMap) -> ParseResult<Vec<CmdToken>, SomeError>;
```
Importantly, the tokeniser does not need to care about whether the token sequence as a whole is valid - it will tokenise `oad3` just fine into `[InsertChild('a'), Delete, Count(3)]` regardless of whether it is a valid command.  This is the only place that `KeyMap` is used.  This type of parsing is quite amenable to macro-ification, and could probably be generated completely from the specification of a macro.

### Phase 2: Parsing command structure into discrete actions

At this point, we have to turn a `Vec<CmdToken>` into some representation of changes that can be applied to `editable_tree::DAG`.  This roughly corresponds to the parsing and code-gen steps in compilers, but simpler due to the restricted nature of commands.

So I think this step corresponds to roughly this specification:
```rust
// This differs from the current `Action` because we are not yet 
// converting `char`s into the corresponding AST nodes.
enum Action {
    InsertChild(char),
    Delete,
    Move(Direction),
    ...
}

// Note that here every action has a count (which will be `1` in most cases)
fn parse(tokens: &[CmdToken]) -> ParseResult<Vec<(usize, Action)>, SomeError>;
```

### Phase 3: Applying changes

Finally, we have to apply the sequence of `Action`s to the AST.  I think this could simply be done with a method on `editable_tree::DAG` like the following (perhaps with special cases for things like quitting Sapling):
```rust
pub fn apply_actions(&mut self, actions: &[(usize, Action)]) -> Result<(), SomeError>;
```
This should make sure that there is only one undo/redo-history entry for each command the user typed (even if it can be composed into multiple `Action`s).

## Pros

- Each phase only handles one thing so there is a clear separation of concerns, making the whole parser easy to reason about.
- This method is highly configurable and extensible.
- The first two stages are completely agnostic of the language that Sapling is editing.  The third phase has to be language-dependent.
- The language-agnosticness means that writing unit tests for each parsing step is very easy.
- `editor::Editor` does not need to know about how the parsing happens specifically.  We can wrap all the parsing into its own module (say, `editor::cmd_parsing`) and then expose a single public function
  ```rust
  pub fn parse_cmd(command: &str, editable_tree: &mut DAG) -> ParseResult<(), SomeError> {
      editable_tree.apply_actions(parse(tokenise(command)?)?);
      ParseResult::Ok(())
  }
  ```
  that parses the command and either:
  - the command is valid: the changes are applied to DAG and `Ok(())` is returned
  - the command was found incomplete in some phase and `ParseResult::Incomplete` is returned
  - the command was found invalid in some phase and `ParseResult::Invalid<SomeError>` is returned
- This tokenise/parse/execute pattern is generally quite well known due to its prevalence in compilers

## Cons
- We still haven't removed the duplication between `Action` and `Command`/`CmdToken`.  To make matters worse, we'll need to add a third enum for keybindings to map to, which adds further duplication.

  We can definitely add a macro to generate `CmdToken` and whatever keybindings get mapped to, but there may be `Action`s which don't have `CmdToken` (like find and replace) so I think having a macro for this would probably become quite messy.  I think that this duplication is probably justified, since `Action` and `CmdToken` have completely different meanings.
- This seems very overkill for 'simple' command parsing.

  I think complexity is necessary to match Vim's complex command system, and I think putting this code into its own module will provide easy opportunity to refactor.  Also, this is I think the only way to satisfy my vision that Sapling should allow both Vim's command structure **and** potentially better ones.

Phew - this turned out to be really long! :sweat_smile:

**Comment by Kixunil** [comment]
So suppose you have a support for `3Jd` as you suggested meaning "delete 3 children" but `3J` is supposed to *also* move cursor down. Then it's essentially the same as `v3jd` in Vim. Activating selection will probably have to be a different action and `d` behaves differently based on mode.

We hit another complexity and I guess each grammar will need to implement its own parser pretty much from scratch anyway.

>  I think that this duplication is probably justified, since `Action` and `CmdToken` have completely different meanings.

I suspected this for a while but wasn't sure if macros could make a difference.

Regarding making the parser language agnostic, I think that we should just keep `char` for the nodes and let each language parse that.

**Comment by kneasle** [comment]
In all honesty, I think the easiest way to remove complexity is to bake Kakoune's selection-based grammar into Sapling.  It's significantly simpler than Vim's, since I'm pretty sure it doesn't let you compose commands into one meta-command - you run each subcommand independently so commands could only be single `CmdToken`s, which would correspond to single `Action`s, and no parsing step is required.

However, I'd quite like Sapling to be as easy to learn as possible for people coming from Vim, and my initial solution for that was to allow the user to switch between Vim's command parser and a better-but-novel grammar inspired by Kakoune.  My preferred would be to have Sapling convert Vim's commands into Kakoune-like commands, since they are simpler.  But I'll agree that some Vim commands like 'v' don't have corresponding Kakoune-style commands.

The more I think about it, the more I'm tempted to just give up on perfectly following Vim's grammar and go with something simpler.  Simplicity also comes with a huge advantage of being easy to reason about and therefore easy to keep working.

**Comment by Kixunil** [comment]
I've never used Kakoune but from my understanding of how you described it, hitting shift during movement is the same as `v`, right? Anyway it looks like all commands are contextual...

One thing that occurred to me is that we probably want to have something like insert mode. We could treat strings as nodes containing chars but hitting `afaoao` just to write `foo` would not bring speed increase. Just like in vim `axy` should switch to insert mode and add nodes `x` and `y`.

Anyway I still fail to see how Kakoune has better grammar except that it might be simpler to implement. Is there a better example?

**Comment by kneasle** [comment]
> I've never used Kakoune but from my understanding of how you described it, hitting shift during movement is the same as `v`, right? Anyway it looks like all commands are contextual...

I also haven't used Kakoune very much, though my friend Charlie Stanton (who inspired this whole project) uses it as his main text editor, and so I've fiddled around with it a bit.

[This blog post](https://cosine.blue/2019-09-06-kakoune.html) is a pretty nice run-down of how Kakoune's grammar works.

One of the main features of Kakoune is that it is *selection-oriented*.  You interact with text primarily by creating as many selections as you need, then performing actions on these selections.  Also having multiple cursors/selections is an awesome feature in general, and one that Kakoune-like editors fully embrace.

> Anyway I still fail to see how Kakoune has better grammar except that it might be simpler to implement. Is there a better example?

The main advantage of this is that it is more interactive and error-friendly than Vim (as well as simply being faster in most cases) - suppose I wanted to find/replace `TestJSON` into `FooJSON` in just the main function of `main.rs`:
```rust
...
use crate::ast::test_json::TestJSON;
...
fn main() {
    // Initialise the logging and startup
    pretty_env_logger::init();
    log::info!("Starting up...");

    // Create an empty arena for Sapling to use
    log::trace!("Creating arena");
    let arena = Arena::new();
    // For the time being, start the editor with some pre-made JSON
    let root = TestJSON::Array(vec![
        TestJSON::True,
        TestJSON::False,
        TestJSON::Object(vec![("value".to_string(), TestJSON::True)]),
    ])
    .add_to_arena(&arena);

    let mut tree = DAG::new(&arena, root);
    let editor = Editor::new(&mut tree, JSONFormat::Pretty, editor::default_keymap());
    editor.run();
}
```

### In Vim

In Vim, I would move my cursor to inside the main function, and then type `va{` to select just the body of that function.  Then I do the substitution: I type `:s/TesTJSON/FooJSON/g<enter>` and I get a message `E486: Pattern not found: TesTJSON` whilst Vim clears the selection I made.  Now I realise that I misspelt `TestJSON` but had no immediate feedback to tell me that this was the case.  So I start again by typing `va{` then 
`:s/TestJSON/FooJSON/g<enter>` which finally works.

### In Kakoune

However, in Kakoune, I move my cursor somewhere inside the main function, and type `<alt-a>{`, which expands the cursor to a single selection containing the body of the main function.  I now type `s` and start typing `TestJSON`.  Kakoune immediately starts showing me where my new selections will end up:

image

Now I get feedback of my typo, and finally finish typing `sTestJSON<enter>`.  Now, I have 5 independent selections, each with their own cursor, so I type `c` to delete all the selections and go into insert mode at each cursor. Now, as I type it start typing from each cursor:

image

So now I finish typing `cFooJSON` and press `<esc>` to go from insert mode to normal mode.  And finally I type `<space>` to get back to a single cursor.

### Comparison

So, in an ideal world where I typed perfectly regardless of the level of editor feedback I would have typed `va{:s/TestJSON/FooJSON/g` (24 keystrokes) in Vim, and `<alt-a>{sTestJSONcFooJSON<esc><space>` (~22) in Kakoune if using `alt` counts as a whole extra keystroke.  So Kakoune is faster even in the case where I typed perfectly.

If I made a mistake, however, the lack of interaction from Vim really makes a difference:  I'd have to type `va{:s/TesTJSON/FooJSON/gva{:s/TestJSON/FooJSON/g` (48 keystrokes, but perhaps better if I used the up arrow after pressing `:` the second time) in Vim, and `<alt-a>{sTesT<backspace>tJSONcFooJSON<esc><space>` (24 keystrokes) in Kakoune.

> One thing that occurred to me is that we probably want to have something like insert mode. We could treat strings as nodes containing chars but hitting `afaoao` just to write `foo` would not bring speed increase. Just like in vim `axy` should switch to insert mode and add nodes `x` and `y`.

Yes - definitely, though I think some experimentation of keybindings would be useful.  I would also really like there to be a way to just start editing a given node as raw text (which feels quite like insert mode to me), and then the tree will be re-parsed either as you type or once you exit that.  This text-style editing shouldn't be the primary way to edit code - it would just be if you wanted something like `x + 1` as an expression and didn't want to insert the tree manually.  It also needs a beefy parsing setup to work properly.

**Comment by Kixunil** [comment]
(Tip: you can reselect last selection in vim by typing `gv` ;) This does **not** mean that selection-oriented editor is useless.)

That's quite interesting. I'm not entirely confident about the benefits but worth a try I think. I consider find/replace a poor replacement (lol) for refactoring (but admittedly, I still use it quite often). So if I were to choose, I'd prefer seeing some refactoring features to having yet another way to find/replace. Note that find/replace doesn't even map to Sapling that well.

> if you wanted something like `x + 1` as an expression and didn't want to insert the tree manually

When I was thinking about AST-based editor idea a long time before I imagined it could work by trying to predict what you're going to type (similarly to pressing tab in bash) and highlight places with missing items using red color.

examples:

`p` -> `pub<RED SPACE>`
`pub(` -> `pub(<RED SPACE>) <RED SPACE>`
`pub(c) <RED SPACE>` -> `pub(crate) <RED SPACE>`
`pub(crate) f` -> `pub(crate) fn <RED SPACE>() { }`
`pub(crate) fn foo(b) { }` -> `pub(crate) fn foo(b: <RED SPACE>) { }`

This could feel more natural than the current approach but hits some problems where the current mode can be more useful. The most significant one is when you need to wrap things.

Ordinary editors do something like this:

Original code:

```rust
foo();
```

What user types:

```rust
loop {
foo();
```

What the editor completes:

```rust
loop {
}
foo();
```

What the user intends:

```rust
loop {
    foo();
}
```

Clearly with AST editor you just select the appropriate sequence of nodes and type some command to wrap it. What's interesting about this operation is some cases might have two options. Contrived example: `foo()` can be turned to `while foo() {}` or `while x { foo(); }`. We need to make this selectable.

At this point I feel like we went too far though, so I wonder if throwing these ideas around makes sense now. If there's not a clear approach on deduplication neither a clear decision how grammar should look like, maybe it's best to wait with deciding on these things at least until we have something somewhat usable so that we can figure out what's the most efficient approach.

Let me end this with a little encouragement: not too long ago I often used `_ =>` in match if I was lazy to type out all the remaining variants, but that allows me to forget stuff, so lately I stopped doing this. Even more crazily, just today I wrote three different methods on enum that match and return appropriate string, but each of those methods behaves the same! Why did I do such duplication? Because those methods are used in different contexts and it's perfectly possible that if I add a new variant in the future, that variant will not behave the same. Having to manually check all places would be worse than this kind of duplication. So let's not worry about duplication too much. It can be a good thing. The bad kind of duplication is one that breaks the "single source of truth" principle.

**Comment by kneasle** [comment]
> Tip: you can reselect last selection in vim by typing `gv`

Wow! I didn't know that :)

> That's quite interesting. I'm not entirely confident about the benefits but worth a try I think. I consider find/replace a poor replacement (lol) for refactoring (but admittedly, I still use it quite often). So if I were to choose, I'd prefer seeing some refactoring features to having yet another way to find/replace. Note that find/replace doesn't even map to Sapling that well.

Multi-cursor/multi-selection editing has other advantages: any time that you want to repeat a whole load of identical operations then it makes a lot of sense to set up a load of cursors in those locations and do them all at the same time.  It also just feels really badass to do something like that.  Even if using multiple cursors isn't the primary way of interacting with an editor, it's a really neat feature to have from the start.

> When I was thinking about AST-based editor idea a long time before I imagined it could work by trying to predict what you're going to type (similarly to pressing tab in bash) and highlight places with missing items using red color.

I still think this would be a cool project (but not really what Sapling is going for).  It'd probably be more easy to implement as a plugin for Vim using tree-sitter to handle the parsing and AST structure detection, rather than building a full new editor ground-up.

> At this point I feel like we went too far though, so I wonder if throwing these ideas around makes sense now. If there's not a clear approach on deduplication neither a clear decision how grammar should look like, maybe it's best to wait with deciding on these things at least until we have something somewhat usable so that we can figure out what's the most efficient approach.

I agree with this - I think the current discussion has definitely been useful, but without an editor to play around with, it's really hard to know what will work well and what won't.  I had mostly forgotten that this whole thing sprang out of trying to remove a relatively small amount of duplication :laughing:.

**Comment by Kixunil** [comment]
Is there anything left to be done before this PR can be merged? I think it's a good start to merge it regardless what we'll do.

**Comment by kneasle** [comment]
Yep - the discussion wasn't really related to the PR :laughing:.  Merging now.

**Pull request merged**
